File: backend/src/models/UngVien.js
const mongoose = require('mongoose');
const { TRANG_THAI_UNG_VIEN } = require('../utils/constants');

const lichSuSchema = new mongoose.Schema({
  trangThaiCu: {
    type: String,
    enum: Object.values(TRANG_THAI_UNG_VIEN)
  },
  trangThaiMoi: {
    type: String,
    enum: Object.values(TRANG_THAI_UNG_VIEN),
    required: true
  },
  nguoiThayDoi: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung'
  },
  ghiChu: {
    type: String,
    trim: true
  },
  thoiGian: {
    type: Date,
    default: Date.now
  }
}, { _id: false });

const goiYSchema = new mongoose.Schema({
  tieuChi: {
    type: String,
    required: true,
    trim: true
  },
  ketQua: {
    type: String,
    enum: ['Dat', 'Khong_dat', 'Chua_danh_gia'],
    default: 'Chua_danh_gia'
  },
  ghiChu: {
    type: String,
    trim: true
  }
}, { _id: false });

const ungVienSchema = new mongoose.Schema({
  hoTen: {
    type: String,
    required: [true, 'Vui lòng nhập họ tên'],
    trim: true,
    maxlength: [100, 'Họ tên không được quá 100 ký tự']
  },
  email: {
    type: String,
    required: [true, 'Vui lòng nhập email'],
    lowercase: true,
    trim: true,
    match: [/^\S+@\S+\.\S+$/, 'Email không hợp lệ']
  },
  soDienThoai: {
    type: String,
    required: [true, 'Vui lòng nhập số điện thoại'],
    trim: true,
    match: [/^[0-9]{10,11}$/, 'Số điện thoại không hợp lệ']
  },
  ngaySinh: {
    type: Date
  },
  gioiTinh: {
    type: String,
    enum: ['Nam', 'Nu', 'Khac']
  },
  diaChi: {
    type: String,
    trim: true
  },
  jobId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'TinTuyenDung',
    required: true
  },
  cvUrl: {
    type: String,
    required: [true, 'Vui lòng upload CV']
  },
  cvFileName: {
    type: String
  },
  thuUngTuyen: {
    type: String,
    trim: true
  },
  kinhNghiem: {
    type: Number,
    default: 0,
    min: 0
  },
  kyNangNoiBat: [{
    type: String,
    trim: true
  }],
  hocVan: {
    type: String,
    trim: true
  },
  trangThai: {
    type: String,
    enum: Object.values(TRANG_THAI_UNG_VIEN),
    default: TRANG_THAI_UNG_VIEN.MOI_NOP
  },
  goiY: {
    type: [goiYSchema],
    default: []
  },
  diemSangLoc: {
    type: Number,
    min: 0,
    max: 100,
    default: null
  },
  nhanXetSangLoc: {
    type: String,
    trim: true
  },
  nguoiSangLoc: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung'
  },
  ngaySangLoc: {
    type: Date
  },
  lichSu: {
    type: [lichSuSchema],
    default: []
  },
  ngayNop: {
    type: Date,
    default: Date.now
  },
  nguonUngTuyen: {
    type: String,
    enum: ['Website', 'Email', 'Gioi_thieu', 'Mang_xa_hoi', 'Khac'],
    default: 'Website'
  }
}, {
  timestamps: true,
  collection: 'ung_vien'
});

// Index
ungVienSchema.index({ jobId: 1 });
ungVienSchema.index({ email: 1, jobId: 1 });
ungVienSchema.index({ trangThai: 1 });
ungVienSchema.index({ ngayNop: -1 });

// Thêm vào lịch sử khi thay đổi trạng thái
ungVienSchema.pre('save', function(next) {
  if (this.isModified('trangThai')) {
    const lichSuMoi = {
      trangThaiCu: this._previousTrangThai || null,
      trangThaiMoi: this.trangThai,
      nguoiThayDoi: this._nguoiThayDoi || null,
      ghiChu: this._ghiChuThayDoi || null,
      thoiGian: new Date()
    };
    
    this.lichSu.push(lichSuMoi);
    
    // Cleanup temporary fields
    delete this._previousTrangThai;
    delete this._nguoiThayDoi;
    delete this._ghiChuThayDoi;
  }
  next();
});

// Method để cập nhật trạng thái với thông tin người thay đổi
ungVienSchema.methods.capNhatTrangThai = function(trangThaiMoi, nguoiThayDoiId, ghiChu = null) {
  this._previousTrangThai = this.trangThai;
  this._nguoiThayDoi = nguoiThayDoiId;
  this._ghiChuThayDoi = ghiChu;
  this.trangThai = trangThaiMoi;
};

const UngVien = mongoose.model('UngVien', ungVienSchema);

module.exports = UngVien;

File: frontend/lib/models/ung_vien.dart
class GoiY {
  final String tieuChi;
  final String ketQua;
  final String? ghiChu;

  GoiY({
    required this.tieuChi,
    required this.ketQua,
    this.ghiChu,
  });

  factory GoiY.fromJson(Map<String, dynamic> json) {
    return GoiY(
      tieuChi: json['tieuChi'] ?? '',
      ketQua: json['ketQua'] ?? 'Chua_danh_gia',
      ghiChu: json['ghiChu'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'tieuChi': tieuChi,
      'ketQua': ketQua,
      'ghiChu': ghiChu,
    };
  }
}

class LichSu {
  final String? trangThaiCu;
  final String trangThaiMoi;
  final String? nguoiThayDoi;
  final String? ghiChu;
  final DateTime thoiGian;

  LichSu({
    this.trangThaiCu,
    required this.trangThaiMoi,
    this.nguoiThayDoi,
    this.ghiChu,
    required this.thoiGian,
  });

  factory LichSu.fromJson(Map<String, dynamic> json) {
    return LichSu(
      trangThaiCu: json['trangThaiCu'],
      trangThaiMoi: json['trangThaiMoi'] ?? '',
      nguoiThayDoi: json['nguoiThayDoi'],
      ghiChu: json['ghiChu'],
      thoiGian: DateTime.parse(json['thoiGian']),
    );
  }
}

class UngVien {
  final String id;
  final String hoTen;
  final String email;
  final String soDienThoai;
  final DateTime? ngaySinh;
  final String? gioiTinh;
  final String? diaChi;
  final String jobId;
  final String cvUrl;
  final String cvFileName;
  final String? thuUngTuyen;
  final int kinhNghiem;
  final List<String> kyNangNoiBat;
  final String? hocVan;
  final String trangThai;
  final List<GoiY> goiY;
  final double? diemSangLoc;
  final String? nhanXetSangLoc;
  final String? nguoiSangLoc;
  final DateTime? ngaySangLoc;
  final List<LichSu> lichSu;
  final DateTime ngayNop;
  final String nguonUngTuyen;
  final DateTime createdAt;
  final DateTime updatedAt;

  UngVien({
    required this.id,
    required this.hoTen,
    required this.email,
    required this.soDienThoai,
    this.ngaySinh,
    this.gioiTinh,
    this.diaChi,
    required this.jobId,
    required this.cvUrl,
    required this.cvFileName,
    this.thuUngTuyen,
    required this.kinhNghiem,
    required this.kyNangNoiBat,
    this.hocVan,
    required this.trangThai,
    required this.goiY,
    this.diemSangLoc,
    this.nhanXetSangLoc,
    this.nguoiSangLoc,
    this.ngaySangLoc,
    required this.lichSu,
    required this.ngayNop,
    required this.nguonUngTuyen,
    required this.createdAt,
    required this.updatedAt,
  });

  factory UngVien.fromJson(Map<String, dynamic> json) {
    return UngVien(
      id: json['_id'] ?? '',
      hoTen: json['hoTen'] ?? '',
      email: json['email'] ?? '',
      soDienThoai: json['soDienThoai'] ?? '',
      ngaySinh: json['ngaySinh'] != null ? DateTime.parse(json['ngaySinh']) : null,
      gioiTinh: json['gioiTinh'],
      diaChi: json['diaChi'],
      jobId: json['jobId'] is String ? json['jobId'] : json['jobId']?['_id'] ?? '',
      cvUrl: json['cvUrl'] ?? '',
      cvFileName: json['cvFileName'] ?? '',
      thuUngTuyen: json['thuUngTuyen'],
      kinhNghiem: json['kinhNghiem'] ?? 0,
      kyNangNoiBat: List<String>.from(json['kyNangNoiBat'] ?? []),
      hocVan: json['hocVan'],
      trangThai: json['trangThai'] ?? 'Moi_nop',
      goiY: (json['goiY'] as List? ?? []).map((e) => GoiY.fromJson(e)).toList(),
      diemSangLoc: json['diemSangLoc']?.toDouble(),
      nhanXetSangLoc: json['nhanXetSangLoc'],
      nguoiSangLoc: json['nguoiSangLoc'],
      ngaySangLoc: json['ngaySangLoc'] != null ? DateTime.parse(json['ngaySangLoc']) : null,
      lichSu: (json['lichSu'] as List? ?? []).map((e) => LichSu.fromJson(e)).toList(),
      ngayNop: DateTime.parse(json['ngayNop'] ?? DateTime.now().toIso8601String()),
      nguonUngTuyen: json['nguonUngTuyen'] ?? 'Website',
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'hoTen': hoTen,
      'email': email,
      'soDienThoai': soDienThoai,
      'ngaySinh': ngaySinh?.toIso8601String(),
      'gioiTinh': gioiTinh,
      'diaChi': diaChi,
      'jobId': jobId,
      'cvUrl': cvUrl,
      'cvFileName': cvFileName,
      'thuUngTuyen': thuUngTuyen,
      'kinhNghiem': kinhNghiem,
      'kyNangNoiBat': kyNangNoiBat,
      'hocVan': hocVan,
      'trangThai': trangThai,
      'goiY': goiY.map((e) => e.toJson()).toList(),
      'diemSangLoc': diemSangLoc,
      'nhanXetSangLoc': nhanXetSangLoc,
      'nguoiSangLoc': nguoiSangLoc,
      'ngaySangLoc': ngaySangLoc?.toIso8601String(),
      'lichSu': lichSu,
      'ngayNop': ngayNop.toIso8601String(),
      'nguonUngTuyen': nguonUngTuyen,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  // Helper methods
  String get trangThaiText {
    const map = {
      'Moi_nop': 'Mới nộp',
      'Dat_sang_loc': 'Đạt sàng lọc',
      'Khong_dat_sang_loc': 'Không đạt',
      'Cho_phong_van': 'Chờ phỏng vấn',
      'Dang_phong_van': 'Đang phỏng vấn',
      'Hoan_thanh_phong_van': 'Hoàn thành PV',
      'Da_tuyen': 'Đã tuyển',
      'Khong_tuyen': 'Không tuyển',
      'Da_nhan_viec': 'Đã nhận việc',
      'Tu_choi': 'Từ chối',
    };
    return map[trangThai] ?? trangThai;
  }
}